### PR TITLE
Require review labels before merging

### PR DESCRIPTION
**Pull request by asteflova**
#### What changes are you introducing?

I propose updating the project's contributing guidelines with a bullet point explaining that each PR must have the `style review done` and `tech review done` labels set before anyone can merge it.

#### Why are you introducing these changes? (Explanation, links to references, issues, etc.)

The idea came up during Red Hat's docs team retrospective when we met to evaluate the process we followed for the Foreman 3.12 release. The issues we discussed and hope to address with this PR include:

* Uncertainty surrounding whether a PR has been properly reviewed and is ready to be merged.
* Accountability about who is responsible for reviewing what.
* Transparency about who has already reviewed a PR and where feedback was provided.

#### Anything else to add? (Considerations, potential downsides, alternative solutions you have explored, etc.)

@Lennonka suggested it would be a good idea to completely disable the ability to merge PRs that don't have these labels. That would be great if GH supports that feature.

#### Checklists

* [x] I am okay with my commits getting squashed when you merge this PR.
* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.12/Katello 4.14 (Satellite 6.16)
* [ ] Foreman 3.11/Katello 4.13
* [ ] Foreman 3.10/Katello 4.12
* [ ] Foreman 3.9/Katello 4.11 (Satellite 6.15; orcharhino 6.8/6.9/6.10)
* [ ] Foreman 3.8/Katello 4.10
* [ ] Foreman 3.7/Katello 4.9 (Satellite 6.14)
* [ ] Foreman 3.6/Katello 4.8
* [ ] Foreman 3.5/Katello 4.7 (Satellite 6.13; orcharhino 6.6/6.7)
* We do not accept PRs for Foreman older than 3.5.

### CODE REVIEW

**Comment by asteflova** [comment]
I'm not sure how to handle PRs that don't need one type of review. For example, a typo that doesn't need tech review. I lean towards setting the label anyway rather than introducing a whole new set of labels (like `tech review not needed`).

**Comment by Lennonka** [comment]
I generally support this, but I'm also not sure how to go about the PRs that don't require SME reviews.

**Comment by ekohl** [comment]
> Uncertainty surrounding whether a PR has been properly reviewed and is ready to be merged.

I think this is a never ending quest and even with years of experience in this area, I can't say I always know well.

Greg has written [a blog](https://theforeman.org/2019/03/merge-time-review-complexity.html) about this quite some time ago. He found review time significantly increases with the number of files touched. That's why I always hammer on keeping PRs small and focused. I can only speak for myself, but that helps me to quicker feel it is properly reviewed and ready to be merged.

> Accountability about who is responsible for reviewing what.

You can map files to people/groups using [code owners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) but I doubt it really helps here. Often you need some SME who shouldn't own the file, but can review the procedure.

In the upstream docs meeting I made sure we always have an agenda item to review PRs that were open too long without activity. That should function as a backstop for PRs that don't receive reviews and allow us (as a team) to make sure someone is assigned.

In short: no hard rules that always work, but develop transparency and procedures that prevent things from going in the wrong direction.

> Transparency about who has already reviewed a PR and where feedback was provided.

I know within Red Hat it's common for some writer to ask an engineer to review. Sometimes that happens outside of GitHub, sometimes within. I would be in favor of always asking them to leave a review on a PR (you don't need any permissions to do so) because it's visible to everyone that it has been reviewed and at which point.

Note it doesn't mean you can't sync up outside of GitHub, but if you do then at least summarize for others what has been discussed and what was concluded.

> @Lennonka suggested it would be a good idea to completely disable the ability to merge PRs that don't have these labels. That would be great if GH supports that feature.

I don't think I've seen that feature. You can build it with actions, but it's not something that exists out of the box.

I'd like to point to https://community.theforeman.org/t/per-package-merge-permissions-in-foreman-packaging-for-plugin-maintainers/39372. It's not the same, but it is about handling permissions around merging and is at least related.

**Comment by asteflova** [comment]
First of all, it's a bit unfortunate that the retrospective happened downstream only; we probably should have considered opening it up but since we didn't know whether we could keep the conversation completely un-confidential and because we had an Agile Practitioner from Red Hat present, downstream-only it was. We can repeat the exercise upstream if needed. However, let me share a few more details about the reasons for this proposal.

> > Uncertainty surrounding whether a PR has been properly reviewed and is ready to be merged.
> 
> I think this is a never ending quest and even with years of experience in this area, I can't say I always know well.
> 
> Greg has written [a blog](https://theforeman.org/2019/03/merge-time-review-complexity.html) about this quite some time ago. He found review time significantly increases with the number of files touched. That's why I always hammer on keeping PRs small and focused. I can only speak for myself, but that helps me to quicker feel it is properly reviewed and ready to be merged.
> 

At the retrospective, we didn't really discuss the length of PRs (although I agree with you). The issue we are hoping to address are situations when a maintainer merges a PR that is still being reviewed. Expecting the author of the PR to set the labels properly would help prevent this: "Don't merge this; I'm not done, I still have feedback to implement."

> > Accountability about who is responsible for reviewing what.
> 
> You can map files to people/groups using [code owners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) but I doubt it really helps here. Often you need some SME who shouldn't own the file, but can review the procedure.
> 
> In the upstream docs meeting I made sure we always have an agenda item to review PRs that were open too long without activity. That should function as a backstop for PRs that don't receive reviews and allow us (as a team) to make sure someone is assigned.
> 
> In short: no hard rules that always work, but develop transparency and procedures that prevent things from going in the wrong direction.

A tech review is not always en engineer's review. For example, when working on a change related to a customer case, it's good to ask the Support team or even that customer for a review. We can't expect them to share their thoughts on GitHub. We can, however, expect the author of a PR to leave a few details about the review and to set the labels accordingly (examples [here](https://github.com/theforeman/foreman-documentation/pull/3258#issuecomment-2329345296) and [here](https://github.com/theforeman/foreman-documentation/pull/3231#issuecomment-2315504177)).

> 
> > Transparency about who has already reviewed a PR and where feedback was provided.
> 
> I know within Red Hat it's common for some writer to ask an engineer to review. Sometimes that happens outside of GitHub, sometimes within. I would be in favor of always asking them to leave a review on a PR (you don't need any permissions to do so) because it's visible to everyone that it has been reviewed and at which point.
> 
> Note it doesn't mean you can't sync up outside of GitHub, but if you do then at least summarize for others what has been discussed and what was concluded.

That would be great. And I suspect the labels can help with it too. When a PR author sets a label but the PR itself doesn't include any record of that feedback, a maintainer can always ask the author to add an explanation.

**Comment by asteflova** [comment]
This whole proposal draws on experiences from other doc teams at Red Hat. By saying that, I'm not suggesting we are trying to impose d/s standards on everyone; only that we want to share what we know about how documentation reviews work on other teams.

The typical workflow for any writer at RH should be: Get SME review, get peer review, merge. I'm hoping that using the labels properly would simply help everyone (both authors and maintainers) be clear on what is expected to happen with each PR.

By the way, right now, the default label set on a new PR is `Not yet reviewed`. Perhaps there should in fact be two labels set by default: `style review needed` and `tech review needed`.

**Comment by Lennonka** [comment]
> By the way, right now, the default label set on a new PR is `Not yet reviewed`. Perhaps there should in fact be two labels set by default: `style review needed` and `tech review needed`.

Amen to that! +1

**Review by apinnick — APPROVED** [review]
LGTM. I think this is a big improvement!

**Comment by ekohl** [comment]
> At the retrospective, we didn't really discuss the length of PRs (although I agree with you). The issue we are hoping to address are situations when a maintainer merges a PR that is still being reviewed. Expecting the author of the PR to set the labels properly would help prevent this: "Don't merge this; I'm not done, I still have feedback to implement."

In GitHub you can enforce that all conversations need to be closed. I haven't used it myself so I have no idea if that helps in practice.

> A tech review is not always en engineer's review. For example, when working on a change related to a customer case, it's good to ask the Support team or even that customer for a review. We can't expect them to share their thoughts on GitHub. We can, however, expect the author of a PR to leave a few details about the review and to set the labels accordingly (examples [here](https://github.com/theforeman/foreman-documentation/pull/3258#issuecomment-2329345296) and [here](https://github.com/theforeman/foreman-documentation/pull/3231#issuecomment-2315504177)).

Absolutely agree and I didn't intend to limit it to an engineer. That's just where my mind went first. Please substitute my engineer for non-writer.

> The typical workflow for any writer at RH should be: Get SME review, get peer review, merge. I'm hoping that using the labels properly would simply help everyone (both authors and maintainers) be clear on what is expected to happen with each PR.

Right now that isn't very transparent to me. When a PR comes in, I often take a look but it would be useful to know. I sometimes try to give an indication how confident I feel about my PRs (on a scale of completely experimental to very confident). I wonder how others think about this.

> By the way, right now, the default label set on a new PR is `Not yet reviewed`. Perhaps there should in fact be two labels set by default: `style review needed` and `tech review needed`.

That shouldn't be hard. We can either get the bot to do it or use GitHub Actions. Shall we spin that off in its own issue?

**Comment by asteflova** [comment]
> > By the way, right now, the default label set on a new PR is `Not yet reviewed`. Perhaps there should in fact be two labels set by default: `style review needed` and `tech review needed`.
> 
> That shouldn't be hard. We can either get the bot to do it or use GitHub Actions. Shall we spin that off in its own issue?

That would be great. I was going to create an issue for https://github.com/theforeman/prprocessor but I don't know if that's the right place.

**Review by maximiliankolb — APPROVED** [review]
LGTM

**Comment by maximiliankolb** [comment]
triage: @Lennonka Are you OK with these changes? Let us know/merge or ping me to merge it.

**Pull request merged**
